feat(popover): add pf-popover

eyevana opened the pull request:
<!-- Labels: feature, needs: AT updates, needs: changelog, ready: branch testing, ready: browser testing, ready: code review, priority: low -->

<!-- Thank you for submitting a pull request! Please note that your pull request *title* MUST adhere to conventional commits to be valid.  For more information: https://www.conventionalcommits.org/en/v1.0.0/ -->

<!-- Reminder: please DO NOT forget your changelog and visual baseline updates! -->

## pf-popover

Implemented Patternfly's [Popover](https://www.patternfly.org/v4/components/popover) as web component.



### Related issues

- (#2291)


### Preview

<!-- Suggest linking to Netlify or a public sandbox; not a resource behind a log-in or VPN -->
Link(s) to demo page(s) where this element can be viewed:
- [Link](https://deploy-preview-2320--patternfly-elements.netlify.app/components/popover/) 


<!-- ### What has changed and why -->

<!-- Summarize files edited as part of this MR along with a brief description of what was changed/why. -->




<!-- ### Testing instructions -->

<!-- Be sure to include detailed instructions on how your update can be tested by another developer. -->


#### Browser requirements

Your component should work in all of the following environments:

- [ ] Latest 2 versions of Edge
- [ ] Internet Explorer 11 (should be useable, not pixel perfect)
- [ ] Latest 2 versions of Firefox (one on Mac OS, one of Windows OS)
- [ ] Firefox 78 (or latest version for Red Hat Enterprise Linux distribution)
- [ ] Latest 2 versions of Chrome (one on Mac OS, one of Windows OS)
- [ ] Latest 2 versions of Safari
- [ ] Android mobile device (such as the Galaxy S9)
- [ ] Apple mobile device (such as the iPhone X)
- [ ] Apple tablet device (such as the iPhone Pro)


### Ready-for-merge Checklist

<!-- Check off items as they are completed.  Feel free to delete items if they are not applicable. -->

- [ ] Expected files: all files in this pull request are related to one request or issue (no stragglers or scope-creep).
- [ ] Tests have been updated to cover these changes.
- [ ] Browser testing passed.
- [ ] Changelog updated.
- [ ] Documentation (README.md, WHY.md, etc.) updated or added.
- [ ] Link to the demo recording: []()
- [ ] Approved by designer.


### Merging

Please **squash** when merging and ensure your commit message uses [conventional commit](https://www.conventionalcommits.org/en/v1.0.0/#summary) formatting.

**Be sure to share your updates with the [patternfly-elements-contribute@redhat.com](mailto:patternfly-elements-contribute@redhat.com) mailing list!**

bennypowers commented on elements/pfe-popover/tsconfig.json:
```diff
@@ -0,0 +1,12 @@
+{
+  "extends": "../../tsconfig.settings.json",
+  "include": ["./**/*"],
```
remove this file

bennypowers commented on tsconfig.settings.json:
```diff
@@ -69,6 +69,9 @@
       "@patternfly/patternfly/*": [
         "@patternfly/patternfly/*"
       ],
+      "@patternfly/pf-popover": [
```
reset this file

bennypowers commented on elements/pfe-popover/test/pfe-popover.e2e.spec.ts:
```diff
@@ -0,0 +1,12 @@
+import { test } from '@playwright/test';
```
this file should be called `pf-popover.e2e.ts`

bennypowers commented on elements/pfe-popover/package.json:
```diff
@@ -0,0 +1,50 @@
+{
+  "name": "@patternfly/pfe-popover",
+  "version": "1.0.0",
```
remove this file

bennypowers commented on elements/pfe-popover/custom-elements-manifest.config.js:
```diff
@@ -0,0 +1,5 @@
+import { pfeCustomElementsManifestConfig } from '@patternfly/pfe-tools/custom-elements-manifest.js';
+
+export default pfeCustomElementsManifestConfig({
```
remove this file

bennypowers commented on elements/pfe-popover/README.md:
````diff
@@ -0,0 +1,50 @@
+Load `<pfe-popover>` via CDN:
+
+```html
+<script src="https://unpkg.com/@patternfly/pfe-popover?module"></script>
````
```suggestion
<script src="https://jspm.dev/@patternfly/elements/pf-popover/pf-popover.js"></script>
```

bennypowers commented on elements/pfe-popover/README.md:
````diff
@@ -0,0 +1,50 @@
+Or, if you are using [NPM](https://npm.im), install it
+
+```bash
+npm install @patternfly/pfe-popover
````
```suggestion
npm install @patternfly/elements
```

bennypowers commented on elements/pfe-popover/README.md:
````diff
@@ -0,0 +1,50 @@
+Then once installed, import it to your application:
+
+```js
+import '@patternfly/pfe-popover';
````
```suggestion
import '@patternfly/elements/pf-popover/pf-popover.js';
```

bennypowers commented on elements/pfe-popover/README.md:
````diff
@@ -0,0 +1,50 @@
+### Basic Tooltip
+
+```html
+<pfe-popover
````
```suggestion
<pf-popover
```
Please rename all files and instances from pfe- to pf-

bennypowers commented on package-lock.json:
```diff
@@ -86,6 +86,15 @@
         "@patternfly/pfe-sass": "^2.0.0-next.5"
       }
     },
+    "elements/pf-popover": {
```
please reset this file

bennypowers commented on elements/pfe-tooltip/BaseTooltip.ts:
```diff
@@ -20,7 +20,6 @@ export abstract class BaseTooltip extends LitElement {
   abstract position?: Placement;
 
   #float = new FloatingDOMController(this, {
-    arrow: true,
```
@brianferry please validate this change

eyevana replied:
@brianferry I was poking at the controller and noticed `arrow` wasn't being used in the code, so I removed it. Let me know if there's something I'm missing.

bennypowers reviewed (changes requested):
time to merge in the main branch. Please address the comments here before merging, to make it easier on you when you merge them in.

Once main is merged and conflicts are resolved, please rename the shadow stylesheets to `.css` and validate their syntax (watch for comments and nesting)

eyevana commented on elements/pfe-popover/pfe-popover.ts:
```diff
@@ -130,50 +124,45 @@ export class PfePopover extends LitElement {
-      info: 'info',
-      success: 'check',
-      warning: 'exclamation',
-      danger: 'dumpster-fire',
```
I had to improvise 🤷🏼‍♀️ 😆

eyevana commented:
@bennypowers thanks for the initial review. I believe I addressed all of your suggestions, but let me know if I've missed anything. I still need to add some test coverage and deets to the PR before moving this out of draft status though. I should have it wrapped up by today.

bennypowers commented:
great. will take a deeper look tomorrow. we're in RC season now so preferably we will keep this PR open until 2.0 final is out

eyevana commented:
@bennypowers I'm ready for another round 👀 when you have time, pleaseee :)

bennypowers commented on elements/pf-popover/pf-popover.ts:
```diff
@@ -174,114 +178,137 @@ export class PopoverShownEvent extends ComposedEvent {
+  @property({ type: Number, reflect: true }) distance?: number = 25;
+  // todo: handle PF4s 'flip' Placement option in flip-behavior
+  // https://www.patternfly.org/v4/components/popover#popover
+  @property({ type: String, reflect: true, attribute: 'flip-behavior' }) flipBehavior?: Placement[] = ['top', 'bottom', 'left', 'right', 'top-start', 'top-end', 'bottom-start', 'bottom-end', 'left-start', 'left-end', 'right-start', 'right-end'];
```
We should avoid accepting and certainly reflecting array props. If this *must* be a list, use a converter like the NumberListConverter

bennypowers commented on core/pfe-core/controllers/floating-dom-controller.ts:
```diff
@@ -108,7 +109,7 @@ export class FloatingDOMController implements ReactiveController {
 
   async #update(placement: Placement = 'top', offset?: Offset) {
-    const { flip, padding, shift } = this.#options;
+    const { flip, padding, shift, fallbackPlacements } = this.#options;
```
These should probably go in show, since they can change on the element

brianferry commented on core/pfe-core/controllers/floating-dom-controller.ts:
```diff
@@ -94,33 +102,52 @@ export class FloatingDOMController implements ReactiveController {
+
+      Object.assign(arrow.style, {
+        left: arrowX != null ? `${arrowX}px` : '',
+        top: arrowY != null ? `${arrowY}px` : '',
```
Noticed a small issue in here where if you set the position to `left` but don't leave enough room it looks like the arrow goes to the middle of the element.  We could add a check like `top: arrowY != null && ['top', 'bottom'].includes(_placement) ? '${arrowY}px' : '',`

brianferry reviewed (commented):
Looking good, made a small update to tooltip w/ the floatingdomcontroller arrow fix & left some suggestions for the css variable and the controller.  Demos look good to me outside of the small edit I mentioned.  Do we need to remove the pfe-sass & pfe-styles folders from this PR?

eyevana commented on elements/pf-popover/pf-popover.css:
```diff
@@ -0,0 +1,179 @@
+  transition: visibility 0s, opacity var(--_animation-speed) cubic-bezier(0.54, 1.5, 0.38, 1.11) 0s;
+  left: 0;
+  top: 0;
+  translate: var(--_floating-content-translate);
```
Yep, should be set on the `container` div...at least that's what I'm seeing on my end.

brianferry replied:
I see what you mean, this is set in the javascript, didn't notice that before.

brianferry commented on elements/pf-popover/pf-popover.ts:
```diff
@@ -0,0 +1,387 @@
+  // todo: alertSeverityScreenReaderText
+  // todo: hideOnOutsideClick
+  // todo: withFocusTrap
+  @property({ type: String, reflect: true }) position: Placement = 'top';
```
Just as a note here, I believe that when the property has a type of `String` you can omit it as it's the default.

eyevana replied:
Cool, cool. I'll clean that up. Thx.

brianferry commented on elements/pf-popover/pf-popover.ts:
```diff
@@ -0,0 +1,387 @@
+  #referenceTrigger?: HTMLElement | null = null;
+
+  #float = new FloatingDOMController(this, {
+    content: () => this.shadowRoot?.querySelector('#popover'),
```
Do we need these query selectors in here or can we use the `_popover` variable?  Same for the arrow, but w/ a @query for the arrow element?

eyevana replied:
Yep, I think you're right!

brianferry commented:
Lookin great, demos all look good and the component is looking super nice!

bennypowers commented on core/pfe-core/controllers/floating-dom-controller.ts:
```diff
@@ -12,22 +11,26 @@ import {
   content: Lazy<HTMLElement>;
   invoker?: Lazy<HTMLElement>;
-  arrow?: boolean;
+  arrow?: Lazy<HTMLElement>;
```
@nikkimk 
Would you please take a look at the api changes here? Do you think they're exposing anything we could cover over? Would they be straightforward to migrate to s different underlying library?

nikkimk replied:
It should be fine.

bennypowers commented on elements/pf-popover/pf-popover.ts:
```diff
@@ -0,0 +1,389 @@
+export class PfPopover extends LitElement {
+  static readonly styles = [styles];
+
+  // todo: alertSeverityScreenReaderText
```
lotsa todos here...

bennypowers commented on elements/pf-popover/pf-popover.ts:
```diff
@@ -0,0 +1,389 @@
+  // todo: alertSeverityScreenReaderText
+  // todo: hideOnOutsideClick
+  // todo: withFocusTrap
+  @property({ reflect: true }) position: Placement = 'top';
```
please add JSDoc blocks for all public props and methods, and separate them by an empty line

bennypowers commented on elements/pf-popover/pf-popover.ts:
```diff
@@ -0,0 +1,389 @@
+  @property({ type: Number, reflect: true }) distance?: number = 25;
+  // todo: handle PF4s 'flip' Placement option in flip-behavior if possible
+  // https://www.patternfly.org/v4/components/popover#popover
+  @property({ attribute: 'flip-behavior', converter: StringListConverter }) flipBehavior?: Placement[] = ['top', 'bottom', 'left', 'right', 'top-start', 'top-end', 'bottom-start', 'bottom-end', 'left-start', 'left-end', 'right-start', 'right-end'];
```
does this have to be a public property? and if so, can the default be applied at the call site with a `??` instead of setting it on the instance?

bennypowers commented on elements/pf-popover/pf-popover.ts:
```diff
@@ -0,0 +1,389 @@
+  // todo: handle PF4s 'flip' Placement option in flip-behavior if possible
+  // https://www.patternfly.org/v4/components/popover#popover
+  @property({ attribute: 'flip-behavior', converter: StringListConverter }) flipBehavior?: Placement[] = ['top', 'bottom', 'left', 'right', 'top-start', 'top-end', 'bottom-start', 'bottom-end', 'left-start', 'left-end', 'right-start', 'right-end'];
+  @property({ type: Boolean, reflect: true, attribute: 'enable-flip' }) enableFlip?: boolean = true;
```
```suggestion
  @property({ type: Boolean, reflect: true, attribute: 'enable-flip' }) enableFlip?: boolean = true;
```

a reflecting boolean attr that defaults to true isn't great because there will be no good way to unset. maybe flip (zing!) the sense to `no-flip`?

bennypowers commented on elements/pf-popover/pf-popover.ts:
```diff
@@ -0,0 +1,389 @@
+  @property({ reflect: true }) footer?: string;
+  @property({ reflect: true }) icon?: string;
+  @property({ reflect: true }) label?: string;
+  @property({ type: Number, reflect: true }) distance?: number = 25;
```
can we make this optional and fallback at the call site with `?? 25` or `options.distance ??= 25`?

bennypowers commented on elements/pf-popover/pf-popover.ts:
```diff
@@ -0,0 +1,389 @@
+  @property({ reflect: true, attribute: 'icon-set' }) iconSet?: string;
+  @property({ type: Boolean, reflect: true, attribute: 'hide-close' }) hideClose?: boolean;
+  @property({ reflect: true, attribute: 'alert-severity' }) alertSeverity?: AlertSeverity;
+  @property({ reflect: true, attribute: 'close-label' }) closeButtonLabel = 'Close popover';
```
likewise here

bennypowers commented on elements/pf-popover/pf-popover.ts:
```diff
@@ -0,0 +1,389 @@
+    invoker: () => this.#referenceTrigger || this._slottedTrigger
+  });
+
+  #slots = new SlotController(this, { slots: [null, 'icon', 'heading', 'body', 'footer'] });
```
```suggestion
  #slots = new SlotController(this, null, 'icon', 'heading', 'body', 'footer');
```
same same but nicer

bennypowers commented on elements/pf-popover/pf-popover.ts:
```diff
@@ -0,0 +1,389 @@
+  connectedCallback() {
+    super.connectedCallback();
+    this.addEventListener('keydown', this._onKeydown);
+    document.addEventListener('click', this._outsideClick);
```
use the `instances` trick here to avoid perf issues

bennypowers commented on elements/pf-popover/pf-popover.ts:
```diff
@@ -0,0 +1,389 @@
+  render() {
+    const { alignment, anchor, styles } = this.#float;
+    return html`
+      <div
+        id="container"
+        style="${styleMap(styles)}"
+        class="${classMap({
+          [anchor]: !!anchor,
+          [alignment]: !!alignment,
+        })}"
+      >
+        <slot id="trigger" @keydown=${this._onKeydown} @click=${this.show}></slot>
+        <dialog
+          id="popover"
+          aria-labelledby="heading"
+          aria-describedby="body"
+          aria-label=${ifDefined(this.label)}
+        > 
+          <div id="arrow"></div>
+          <div id="content" part="content">
+            ${this._renderCloseButton()} ${this._renderHeader()} ${this._renderBody()} ${this._renderFooter()}
+          </div>
+      </dialog>
+      </div>
```
```suggestion
      <div id="container"
            style="${styleMap(styles)}"
            class="${classMap({
              [anchor]: !!anchor,
              [alignment]: !!alignment,
            })}">
        <slot id="trigger" @keydown=${this._onKeydown} @click=${this.show}></slot>
        <dialog id="popover"
                aria-labelledby="heading"
                aria-describedby="body"
                aria-label=${ifDefined(this.label)}> 
          <div id="arrow"></div>
          <div id="content" part="content">
            ${this._renderCloseButton()}
            ${this._renderHeader()}
            ${this._renderBody()}
            ${this._renderFooter()}
          </div>
        </dialog>
      </div>
```
just formatting nits

bennypowers commented on elements/pf-popover/pf-popover.ts:
```diff
@@ -0,0 +1,389 @@
+    }
+  }
+
+  protected _renderCloseButton() {
```
:thinking: can we inline these? we've tended to avoid breaking up the templates into private render methods, unless we had a compelling reason

eyevana commented:
@bennypowers 😣 I tried to make focus trapping work by using the dialog's `showModal` method & overriding browser styling to make it look like a regular popover, but the browser renders the dialog in the `#top-layer`, so I'm not able to re-position it by its trigger element. https://deploy-preview-2320--patternfly-elements.netlify.app/components/popover/demo/ <--- see `focus-trap` demo. Do any other alternatives come to mind?

eyevana commented on elements/pf-popover/pf-popover.ts:
```diff
@@ -450,7 +442,7 @@ export class PfPopover extends LitElement {
       fallbackPlacements: this.flipBehavior,
     });
-    this._popover?.show();
+    this.focusTrap ? this._popover?.showModal() : this._popover?.show();
```
@bennypowers I actually removed that bit of code in my last commit. My thought was to use `showModal` in the `focusTrap` scenario instead of `show` and then just override the styling so that it would look identical to a non-focus trapped popover.

bennypowers replied:
yeah that sounds about right

eyevana commented:
@bennypowers read through the [explainer](https://open-ui.org/components/popover.research.explainer/). If we were to switch, the only part I'm concerned about so far is the positioning of the popover to the trigger since the popover would be added to the top layer (similar to the dialog). Seems like we could user [anchoring](https://open-ui.org/components/popover.research.explainer/#anchoring) to achieve this, but it's still [experimental](https://developer.chrome.com/blog/tether-elements-to-each-other-with-css-anchor-positioning/). This would replace our need for the floating-ui dependency though 🥳 I just need to keep reading to see if the API we expose now for positioning (`flip-behavior`, `no-flip`, etc...) would still fly.

eyevana commented:
@bennypowers after digging into [this](https://developer.chrome.com/blog/tether-elements-to-each-other-with-css-anchor-positioning/#position-fallback-and-auto-positioning) some more, I'd feel more comfortable pulling out the property for `flip-behavior`, where we allow the dev to provide an array of fallback placements for the popover if the initial position gets clipped...at least until we move to the popover API + anchor CSS positioning. I see how we could eventually expose an explicit list of fallbacks, but not sure how we could allow the dev to define the list themselves. What do you think?

bennypowers commented on elements/pf-popover/pf-popover.ts:
```diff
@@ -54,10 +54,11 @@ export class PopoverShownEvent extends ComposedEvent {
- *
  * A Popover displays content in a non-modal dialog and adds contextual information or provides resources via text and links.
  *
+ * Unlike the Patternfly React pattern, this component does not trap focus in the popover dialog.
```
I think we should move this to the docs site contents, as the content of `renderOverview`

eyevana replied:
Gotcha. On it!

bennypowers commented on elements/pf-popover/docs/pf-popover.md:
```diff
@@ -6,6 +6,8 @@
 
 {% endrenderOverview %}
 
+<p>Unlike the Patternfly React implementation, this component does not trap focus in the popover dialog. If you woud like to trap focus, consider using a modal dialog instead.</p>
```
```suggestion
Unlike the Patternfly React implementation, this component does not trap focus in the popover dialog. If you woud like to trap focus, consider using a modal dialog instead.
```

eyevana commented on elements/pf-popover/pf-popover.ts:
```diff
@@ -306,9 +306,9 @@ export class PfPopover extends LitElement {
-    const screenReaderText = this.alertSeverity ?
-      html`<span class="sr-only">${this.alertSeverityText ?? `${this.alertSeverity} alert:`}</span>`
-      : nothing;
+    const screenReaderText = !this.alertSeverity ? nothing : html`
```
@bennypowers saw a similar change in a previous commit...why are we flipping the condition around? TI(will)L? :)

bennypowers replied:
it's nice to have the big html chunk after wards, because then there's no trailing "or empty" afterwards.

in other words, since in cases like this we mean to say "only show this template if", it makes sense to put all the sad-path stuff together

eyevana commented:
@bennypowers the demo page looks soooo much better. Thank you! 🎉
